### DISCUSSION TITLE
Connect ELRS RX to ELRS TX - Telemetry missing

### DISCUSSION THREAD

**Discussion by Target0815** [ExpressLRS/ExpressLRS#2279] [Q&A]
Hi,

I would like to connect an ELRS RX with an ELRS TX for a relay setup. This is possible by converting from serial full duplex to half duplex.

Technically this is no problem and works. The problem is that RC data is transmitted, but no telemetry.

In Discord I got a tip that the CRSF packet data for ELRS RX and ELRS TX might be different. However, I have not found much on this subject.

Can someone explain me how it is with the CRSF packet data for RX and TX? Maybe @CapnBry has a little bit of time for this?

Thanks.

**Comment by CapnBry** [comment]
I would say the problem is most likely in the conversion from full duplex to half duplex?

`handset <-> gateway RX <-> gateway TX <-> model`

After any packet is sent from gateway RX, the half duplex connection must immediately be flipped to receive to get the data. You can verify by making sure you're getting opentxsync packets from the gateway TX even when there's no model connected.

The format of the telemetry is not modified receiving vs transmitting, although the destination address may be changed but (without looking it up) I believe the RX code supports receiving 0xC8 or 0xEE just fine?

**Comment by Target0815** [comment]
@CapnBry
 
Hi Bryan, thanks for your reply.

I had first tried to make a conversion full to half duplex with a diode or resistor. That works too, unfortunately only for RC. Then I noticed that this solution is a one way street and the telemetry does not find its way back.

Then I tested the bidirectional full to half duplex converter in the attachment. However, it does not work with that either. I take it from your answer that I guess only hardware can't solve the problem? So the immediate switch to receive data?

What would you recommend as an alternative? RP2040 / ESP32 with appropriate programming? If the format of the telemetry is the same on both sides, this should be feasible. Then you could also send the LQ value from the gateway TX <-> model to the handset.

The problem for me is, that I can handle the programming of the serial interface, but the handling of the CRSF frames is new territory for me. There I would need probably a little support ...

[Datasheet_FullDuplexHalfDuplex_V1.1.pdf](https://github.com/ExpressLRS/ExpressLRS/files/11909188/Datasheet_FullDuplexHalfDuplex_V1.1.pdf)

**Comment by CapnBry** [comment]
I do not know if that converter is suitable for this application, I've never used one before. You'd have to do your own testing to see if you're getting anything at `gateway RX` on its receive pin. This can be tested with any sort of test device hooked to it. I'd recommend compiling yourself some code that just opens a UART at 420000 and dumps any data it gets to Serial.

LinkStats will never be passed back from `gateway TX` to your handset-- LinkStatistics is not considered an external telemetry item which would be passed on. LinkStats are generated by the receiving device as a special operation and you'd really have to make changes in many places to prevent that behavior.

**Reply by Target0815** [reply]
Ok, the UART handling I will still test. But it seems to me almost easier to take a separate RP2040 for the purpose of conversion.

To the LinkStats: with the normal use of the TX module in the handheld transmitter, the data (LQ, RSSI) are also transferred to OpenTX/EdgeTX. In principle, this should then also go via relay. But ok, I will have to test it ...

First of all many thanks for the information. That makes a lot clearer.

**Comment by CapnBry** [comment]
But that's not how our telemetry system works. It isn't a straight passthrough connection from the FC to the handset, therefore LinkStatistics won't just "pass through" from `gateway RX` to the handset. Yes, `gateway TX` will send it to `gateway RX` but it will just be discarded as an unknown packet. 

After I looked at the code this morning, I also think there are some changes needed for telemetry to pass through at all because the RX code will only recognize packets with a CRSF destination of `CRSF_ADDRESS_CRSF_RECEIVER` or `CRSF_ADDRESS_FLIGHT_CONTROLLER`, and when they come from gatewayTX I believe it will have a destination of `CRSF_ADDRESS_RADIO_TRANSMITTER`.

**Reply by Target0815** [reply]
Then it seems not as easy as I thought. To better understand why I want to make: in cold weather I want to sit in the car and also sometimes fly 10 km or more. But this does not work if I have the TX module in the car. So a relay is necessary.

Let's assume a CRSF_ADDRESS_RADIO_TRANSMITTER would exist. Then this would be an additional mode "relay" of ELRS, similar to AirPort. I would probably like to have and express as a wish, but I am well aware that this is not just so ...

May I send you a private message about this?

**Comment by CapnBry** [comment]
I'm pretty sure no new mode would have to be added, just some minor stuff for accepting the telem, some not-so-big changes to support half duplex RXes (at least in this form), then some major restructuring to be allow incoming linkstats to override the internal linkstats in a way that doesn't upset me. That's when you'll find the 800 other things that need to be addressed probably. Now is also not the time for me to work on this when there's more pressing things on the agenda.

Aren't we messaging right now? :-D Actually this is about as far as I'm going on this journey at this time.

**Reply by Target0815** [reply]
Ok and thank you for the infos.

**Comment by CapnBry** [comment]
Oof this took me like an hour and half to get a test setup running, but I've made a PR that stops rejecting the telemetry received at `gateway RX` and I get telemetry relayed back using a DUPLETX ESP as the `gateway TX`.
https://github.com/ExpressLRS/ExpressLRS/pull/2287

I get a ton of "sensor lost" messages at the handset as the rate, and none of the Lua receiver(s) configuration seems to work either (they never show up), but maybe that's a better starting point than outright rejecting the data.

**Comment by Target0815** [comment]
Maybe the PR is already enough so that we can use the relay to some extent ... I would like to test. What exactly do we need as hardware for the gateway TX?

**Comment by CapnBry** [comment]
The TX is unmodified, it is just the RX that needs updating. I tested it with a full-duplex TX because de-half-duplex-inverting a regular TX module was not an exercise I wanted to attempt to debug. If telemetry from a TX arrives at an updated RX it will now forward it, without the update it will silently discard it. So if this doesn't start forwarding telemetry for you, you're going to need to get a logic analyzer / oscilloscope on the RX's RX pin to figure out where to go next.

**Comment by Target0815** [comment]
I could do tests with Happymodel EP1 (DUAL) TCXO receivers if someone will make me a firmware for it. Logic analyzer and Oscilloscope are also available ...

**Comment by CapnBry** [comment]
Use the Configurator.  But not right now because platformio just broke everyone's builds around the world for all projects with their forced update, you will not be able to compile anything from any project right now. 😡

**Comment by Target0815** [comment]
The first test was successful! Telemetry went now immediately, although I still have to see how coherent the values are. In the next days I will do a flight test.

About the binding phrase: here it makes sense to use two binding phrases? One for Handheld -> Relay RX, a second or the normally used one for Relay TX -> Plane?

**Comment by CapnBry** [comment]
Nice! I found that I'd get a lot of sensor lost messages on the in-car handset so you might have to tweak the TLM Ratios on each TX to find just the right combination that keeps the telemetry pipeline full without wasting too much bandwidth on it.

Oh yes 100% you need to use 2x binding phrases. Two receivers on the same binding phrase will connect to one TX, and if there are two TXes it will be random which each will connect to and it could switch mid flight. Model Match isn't designed for that either (in case that was the next question).

**Comment by Target0815** [comment]
Yes, I was actually wondering whether Binding Phrase or Model Match was the right one. Because I have so far only used the Binding Phrase and that works great, I then took a second phrase.

Sensor lost: I once had the problem that only 10 parameters were updated (all from the TX in the handset). I just restarted the RX and it then ran normally. I must still observe whether something hangs up there.

I currently use 100 Hz full and 1:4 telemetry. This is also my usual setting.

**Comment by CapnBry** [comment]
Regarding all the sensors not coming, if a telemetry item is in the middle of transferring (which is constant) and the connection on one end is restarted, then it can take a while to time out before resuming sending telemetry. That means sometimes it may take a bit to resync. This usually takes 20s or so if there's only one TX/RX pair and I think it shouuuuuuuld be the same even with a gateway, so if you see that happen again try waiting a minute to see if resolves itself. It is only a problem on startup from what I've seen.

**Comment by Target0815** [comment]
I received various new hardware today, including RM Boxer ELRS and RM Ranger TX. Boxer connects to an RM EP1. Both are flashed with 3.x.x-maintenance. I assume above PR is included there.

Now when I connect the RM Ranger TX to the EP1, the Ranger reports "No handset". Neither RC nor telemetry is transmitted.

If I connect a Full Crossfire TX instead of Ranger, it immediately reports CRSF V2 as input protocol. RC and telemetry work with it. Therefore the conversion Full Duplex to Half Duplex works.

Furthermore I noticed that via ELRS LUA script on the Boxer handset the menu item "Other Devices" is missing, if the converter (resistor) is connected.

Currently I only have a 1k resistor on the RM EP1 from the TX output to the RX input and from there a connection to the S.Port of the Ranger TX.

Currently I can not explain why it worked with the PR and now with 3.x.x-maintenance no longer ...

**Comment by Target0815** [comment]
> Currently I can not explain why it worked with the PR and now with 3.x.x-maintenance no longer ...

The problem could be solved ... The baudrate of Relay-TX (400.000) and Relay-RX (420.000) did not match. I have now set the RX to 400.000 baud. 

The baud rate of the TX can not be changed? Sure, if the TX is in the handset, but not in the WebUI and not on the hardware page (there are only options for the backpack).

I have a question about the baud rates: why are they different for TX and RX?

**Comment by CapnBry** [comment]
The TX autobauds to any supported baudrate (400000, 115200, 5250000, 3750000, 1870000, 921600, 2250000). It can not autobaud to 420k because it is too close to 400k and as you found, if it picks the wrong one it doesn't work at all.

They are different bauds because that's what TBS made it.

**Comment by StiOII** [comment]
Will be a relay option officially added to elrs?

**Comment by Target0815** [comment]
The change described above is included in the 3.3.0 release. So I guess you can call this official.

**Comment by StiOII** [comment]
Oh cool!
I'm still missing - does a relay with telemetry is fully supported now?
is there any setup guide?

Can be a relay Tx also controlled with the Lua script?

**Comment by Target0815** [comment]
You can simply answer the questions yourself if you read through the above discussion ... The point was that the telemetry did not work before.

There is no setup guide because the setup is simple: two ELRS routes with different binding phrase. Set the RX in the relay to 400 kBaud, TX line from it via 1k resistor to RX line, from there to pin 5 of the relay TX.

The TX in the relay can be set via LUA script, if you plug it shortly to a suitable transmitter.

**Reply by hiarmen** [reply]
What is a pin 5? I only have 4 pins: GND/5V/Tx/Rx. A simple hand drawing would be appreciated on:  TX line from it via 1k resistor to RX line, from there to pin 5 of the relay TX. Thanks!

**Reply by Target0815** [reply]
Pin 5 of the TX module ...

**Reply by hiarmen** [reply]
I see. Instead of using Tx module I am trying to setup a repeater with 2 Rx modules, one of which flashed Rx-as-Tx. Hence no 5th pin. Connected receiver to [Rx-as-Tx] with rx=>tx and tx=1k=>rx, with baud 400k on both. Will this work at all?

**Comment by StiOII** [comment]
Thank you. It's was too easy to believe )
I thought it can't work this way because second tx will not accept packets from rx due to wrong id, something like this: CRSF_ADDRESS_FLIGHT_CONTROLLER = (0xC8)

For the Lua script I meant to control second tx without pluging to a radio of course.

**Comment by Target0815** [comment]
In the constellation as described above, RC has always run in the relay. The only new thing is that the telemetry now also finds its way back to the transmitter, which is why I started this discussion.

What you meant with the LUA script is already clear to me. But for LR you set it once and then you don't have to do it again.

**Reply by iport-snk** [reply]
have you finally got your retransmitter working? I also would like to implement the same schema

**Comment by CapnBry** [comment]
It is "officially supported" currently, but takes some work to get set up and doesn't provide all features.

Even though it works but it isn't great so I don't really recommend it. It is one of the things you can do with ExpressLRS but maybe shouldn't. The telemetry is significantly delayed, more than doubling the amount of latency in receiving telemetry items, which already isn't great due to our architecture. Link telemetry also shows just the actual link, not the relay link information. You can not control the remote TX module from the Lua and likely will never be able to. This is because they both have the same "address" on the CRSF network. I'm not sure if using scripts for configuring the flight controller on the handset will work either. 

Some of this could be addressed but it will never be a great solution.

**Reply by shheyman** [reply]
I’m interested in diving into this and expanding on this feature and the idea of en ELRS network. I’ve been looking at the code but can’t seem to find where the half-duplex implementation on a single pin is handled by the ESP32 in the TX module. Could you point me to that please if you have a moment?

**Reply by iport-snk** [reply]
I guess you are looking for 
CRSFHandset::duplex_set_TX() 
CRSFHandset::duplex_set_RX()
Can you give more details about architecture you are implementing?

**Comment by StiOII** [comment]
Thanks for clarifying Capnbry!
That's why I've asked about the official support, meaning that it all will work flawlessly. So basically it is not, as the problems you've described are significant.

Any chance that the great solution can be ever found in the future?
Or it's never will be possible with the current software concept?

**Comment by agoliveira** [comment]
Hello

Following a suggestion from Target0815, I would like to add that I tested this setup yesterday, and worked like a charm. Thank you all :)
I put everything in a box so it is a matter of just adding your TX module and going for it. The increased latency isn't an issue for me as this will be used for long-range flights, mostly wings.
The only thing I would like to see is the correct link telemetry (ie, from the airplane, not the receiver in the box) but it is not a big issue.

**Reply by EUSKR** [reply]
Would you mind sharing the final schematic of your setup? Trying to make a similar one work.

**Comment by flacon2000** [comment]
@Target0815  Telemetry works, but I encountered the following problem: telemetry from the relay overlaps telemetry from the drone. I see the RSS from relay, but I need it from a drone. At the same time, I receive RxBAT from the drone and other nonovelapping telemetry.

**Comment by Target0815** [comment]
> I see the RSS from relay, but I need it from a drone.

It is already written above that this is the case ... And that more programming is required for this.

**Comment by Maxeemo** [comment]
Hi everyone, I'm trying to make a bridge starting from an elrs module connected to my radio and the related elrs rx connected to the Crossfire TX Full on my ground station, searching online I found this discussion... Honestly I didn't understand how to connect the tx and rx pins of the elrs receiver to the Crossfire TX module, can you help me?
